pkg/model/runtime.go

```diff
@@ -0,0 +1,201 @@
+type RuntimeModel struct {
+	backstageDeployment *BackstageDeployment
+	backstageService    *BackstageService
+
+	localDbStatefulSet *DbStatefulSet
+	localDbService     *DbService
+	localDbSecret      *DbSecret
+
+	Objects []BackstageObject
+}
+
+// Registers config object
+func registerConfig(key string, factory ObjectFactory, need needType) {
+	runtimeConfig = append(runtimeConfig, ObjectConfig{Key: key, ObjectFactory: factory, need: need})
+}
+
+// InitObjects performs a main loop for configuring and making the array of objects to reconsile
+func InitObjects(ctx context.Context, backstageMeta bsv1alpha1.Backstage, backstageSpec *DetailedBackstageSpec, ownsRuntime bool, isOpenshift bool) (*RuntimeModel, error) {
+
+	// 3 phases of Backstage configuration:
+	// 1- load from Operator defaults, modify metadata (labels, selectors..) and namespace as needed
+	// 2- overlay some/all objects with Backstage.spec.rawRuntimeConfig CM
+	// 3- override some parameters defined in Backstage.spec.application
+	// At the end there should be an array of runtime Objects to apply (order optimized)
+
+	lg := log.FromContext(ctx)
+
+	//objectList := make([]BackstageObject, 0)
+	model := &RuntimeModel{Objects: make([]BackstageObject, 0)}
+
+	// looping through the registered runtimeConfig objects initializing the model
+	for _, conf := range runtimeConfig {
+
+		// creating the instance of backstageObject
+		backstageObject := conf.ObjectFactory.newBackstageObject()
+		var defaultErr error
+		var overlayErr error
+
+		// reading default configuration defined in the default-config/[key] file
+		// mounted from the 'default-config' configMap
+		// this is a cluster scope configuration applying to every Backstage CR by default
+		if err := utils.ReadYamlFile(utils.DefFile(conf.Key), backstageObject.Object()); err != nil {
+			defaultErr = fmt.Errorf("failed to read default value for the key %s, reason: %s", conf.Key, err)
+			//lg.V(1).Info("failed reading default config", "error", err.Error())
+		}
+
+		// reading configuration defined in BackstageCR.Spec.RawConfigContent ConfigMap
+		// if present, backstageObject's default configuration will be overriden
+		overlay, overlayExist := backstageSpec.RawConfigContent[conf.Key]
+		if overlayExist {
+			if err := utils.ReadYaml([]byte(overlay), backstageObject.Object()); err != nil {
+				overlayErr = fmt.Errorf("failed to read overlay value for the key %s, reason: %s", conf.Key, err)
+			}
+		}
+
+		// throw the error if raw configuration exists and is invalid
+		// throw the error if there is invalid or no configuration (default|raw) for Mandatory object
+		// continue if there is invalid or no configuration (default|raw) for Optional object
+		// TODO separate the case when configuration does not exist (intentionally) from invalid configuration
+		if overlayErr != nil || (!overlayExist && defaultErr != nil) {
+			if conf.need == Mandatory || (conf.need == ForLocalDatabase && *backstageSpec.Database.EnableLocalDb) {
```

Comment:
Just to be safe, could we use the `LocalDbEnabled()` method you introduced instead?
```suggestion
			if conf.need == Mandatory || (conf.need == ForLocalDatabase && backstageSpec.LocalDbEnabled()) {
```

Reply:
I think so. +1